compiler bug caused by adding str() string as a member of an interface

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: macos, macOS, 12.0.1, 21A559
Processor: 10 cpus, 64bit, little endian, Apple M1 Max
CC version: Apple clang version 13.0.0 (clang-1300.0.29.3)

getwd: /Users/thomas
vmodules: /Users/thomas/.vmodules
vroot: /Users/thomas/Unix/v-latest
vexe: /Users/thomas/Unix/v-latest/v
vexe mtime: 2021-11-22 00:13:43
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 24ba660.e77a110

Git version: git version 2.34.0
Git vroot status: weekly.2021.46-64-ge77a1100
.git/config present: true
thirdparty/tcc status: thirdparty-macos-amd64 689c8a02

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg bug3.v`
```v
module main

interface Str {
	str() string
}

struct St {}

fn (s St) str () string {
	return "s"
}

fn printer (s Str) {
	println('$s')
}

fn main () {
	s := St{}
	printer(s)
}
```

**What did you expect to see?**

the function is called and prints s.


**What did you see instead?**
```
==================
                               ^~~~~~~~~
/tmp/v_501/bug3.8850662568002461293.tmp.c:10697:69: error: implicit declaration of function 'main__Str_str' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        println( str_intp(2, _MOV((StrIntpData[]){{_SLIT0, 0xfe10, {.d_s = main__Str_str(s)}}, {_SLIT0, 0, { .d_c = 0 }}})));
                                                                           ^
/tmp/v_501/bug3.8850662568002461293.tmp.c:10697:69: note: did you mean 'main__St_str'?
/tmp/v_501/bug3.8850662568002461293.tmp.c:10691:24: note: 'main__St_str' declared here
VV_LOCAL_SYMBOL string main__St_str(main__St s) {
                       ^
/tmp/v_501/bug3.8850662568002461293.tmp.c:10697:69: warning: incompatible integer to pointer conversion initializing 'byte *' (aka 'unsigned char *') with an expression of type 'int' [-Wint-conversion]
        println( str_intp(2, _MOV((StrIntpData[]){{_SLIT0, 0xfe10, {.d_s = main__Str_str(s)}}, {_SLIT0, 0, { .d_c = 0 }}})));
                                                                           ^~~~~~~~~~~~~~~~
2 warnings and 1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -7295,13 +7295,13 @@ static inline $interface_name I_${cctype}_to_Interface_${interface_name}($cctype
 			if g.pref.build_mode != .build_module {
 				methods_struct.writeln('\t{')
 			}
-			if st == ast.voidptr_type {
-				for mname, _ in methodidx {
-					if g.pref.build_mode != .build_module {
-						methods_struct.writeln('\t\t._method_${c_name(mname)} = (void*) 0,')
-					}
-				}
-			}
+			// if st == ast.voidptr_type {
+			// 	for mname, _ in methodidx {
+			// 		if g.pref.build_mode != .build_module {
+			// 			methods_struct.writeln('\t\t._method_${c_name(mname)} = (void*) 0,')
+			// 		}
+			// 	}
+			// }
 			mut methods := st_sym.methods
 			match st_sym.info {
 				ast.Struct, ast.Interface, ast.SumType {

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -7295,13 +7295,13 @@ static inline $interface_name I_${cctype}_to_Interface_${interface_name}($cctype
 			if g.pref.build_mode != .build_module {
 				methods_struct.writeln('\t{')
 			}
-			// if st == ast.voidptr_type {
-			// 	for mname, _ in methodidx {
-			// 		if g.pref.build_mode != .build_module {
-			// 			methods_struct.writeln('\t\t._method_${c_name(mname)} = (void*) 0,')
-			// 		}
-			// 	}
-			// }
+			if st == ast.voidptr_type {
+				for mname, _ in methodidx {
+					if g.pref.build_mode != .build_module {
+						methods_struct.writeln('\t\t._method_${c_name(mname)} = (void*) 0,')
+					}
+				}
+			}
 			mut methods := st_sym.methods
 			match st_sym.info {
 				ast.Struct, ast.Interface, ast.SumType {
@@ -7367,7 +7367,7 @@ static inline $interface_name I_${cctype}_to_Interface_${interface_name}($cctype
 					// .speak = Cat_speak_Interface_Animal_method_wrapper
 					method_call += iwpostfix
 				}
-				if g.pref.build_mode != .build_module {
+				if g.pref.build_mode != .build_module && st != ast.voidptr_type {
 					methods_struct.writeln('\t\t._method_${c_name(method.name)} = (void*) $method_call,')
 				}
 			}

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1141,6 +1141,19 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 		}
 		if typ != ast.string_type {
 			expr := node.args[0].expr
+			typ_sym := g.table.get_type_symbol(typ)
+			if typ_sym.kind == .interface_ && (typ_sym.info as ast.Interface).defines_method('str') {
+				g.write('${c_name(print_method)}(')
+				rec_type_name := util.no_dots(g.cc_type(typ, false))
+				g.write('${c_name(rec_type_name)}_name_table[')
+				g.expr(expr)
+				dot := if typ.is_ptr() { '->' } else { '.' }
+				g.write('${dot}_typ]._method_str(')
+				g.expr(expr)
+				g.write('${dot}_object')
+				g.writeln('));')
+				return
+			}
 			if g.is_autofree && !typ.has_flag(.optional) {
 				// Create a temporary variable so that the value can be freed
 				tmp := g.new_tmp_var()

--- a/vlib/v/gen/c/str_intp.v
+++ b/vlib/v/gen/c/str_intp.v
@@ -125,6 +125,7 @@ fn (mut g Gen) str_val(node ast.StringInterLiteral, i int) {
 	expr := node.exprs[i]
 
 	typ := g.unwrap_generic(node.expr_types[i])
+	typ_sym := g.table.get_type_symbol(typ)
 	if typ == ast.string_type {
 		if g.inside_vweb_tmpl {
 			g.write('vweb__filter(')
@@ -139,6 +140,15 @@ fn (mut g Gen) str_val(node ast.StringInterLiteral, i int) {
 			}
 			g.expr(expr)
 		}
+	} else if typ_sym.kind == .interface_ && (typ_sym.info as ast.Interface).defines_method('str') {
+		rec_type_name := util.no_dots(g.cc_type(typ, false))
+		g.write('${c_name(rec_type_name)}_name_table[')
+		g.expr(expr)
+		dot := if typ.is_ptr() { '->' } else { '.' }
+		g.write('${dot}_typ]._method_str(')
+		g.expr(expr)
+		g.write('${dot}_object')
+		g.write(')')
 	} else if node.fmts[i] == `s` || typ.has_flag(.variadic) {
 		mut exp_typ := typ
 		if expr is ast.Ident && g.comptime_var_type_map.len > 0 {

--- a/vlib/v/tests/interface_str_method_test.v
+++ b/vlib/v/tests/interface_str_method_test.v
@@ -1,0 +1,20 @@
+interface Str {
+	str() string
+}
+
+struct St {}
+
+fn (s St) str() string {
+	return 's'
+}
+
+fn printer(s Str) string {
+	println(s)
+	return '$s'
+}
+
+fn test_interface_str_method() {
+	s := St{}
+	ret := printer(s)
+	assert ret == 's'
+}

